src/main/java/components/Storage.java

```diff
@@ -0,0 +1,246 @@
+
+            // Create the directory if it doesn't exist
+            if (!parentDir.exists()) {
+                boolean dirCreated = parentDir.mkdirs();
+                if (!dirCreated) {
+                    throw new DukeException("Failed to create directory at " + parentDir.getAbsolutePath());
+                }
+            }
+
+            // Create a File object for the file to be created
+            File file = new File(this.FILEPATH);
+
+            // Create the file if it doesn't exist
+            if (!file.exists()) {
+                boolean fileCreated = file.createNewFile();
+                if (!fileCreated) {
+                    throw new DukeException("Failed to create store file at " + file.getAbsolutePath());
+                }
+            }
+        
+            //it is guaranteed that the file exists
+            store = file;
+        } catch (IOException e) {
+               throw new DukeException(FILEPATH + " cannot be created.");
+        }
+    }
+
+    /**
+     * Reads data from the file and returns a TaskList object
+     * @param storage
+     * @return
+     * @throws DukeException
+     */
+    public TaskList readData(Storage storage) throws DukeException {
+        BufferedReader br;
+        TaskList result = new TaskList();
+        try {
+            br = new BufferedReader(new FileReader(this.store));
+            String line;
+            while ((line = br.readLine()) != null) {
+                //[T][X] read book 
+                if (line.startsWith("[T]")) {
+                    result.addTask(new ToDo(
+                        line.substring(7)),
+                        storage,
+                        true
+                        );
+                    if (line.charAt(4) == 'X') {
+                        result.get(result.size() - 1).markAsDone();
+                    }
+                //[D][ ] return book (by: Sunday)
+                } else if (line.startsWith("[D]")) {
+                    int byIndex = line.indexOf("(by: ");
+                    result.addTask(new Deadline(
+                        line.substring(7, byIndex - 1), 
+                        line.substring(byIndex + 5, line.length() - 1)),
+                        storage,
+                        true
+                        );
+                    if (line.charAt(4) == 'X') {
+                        result.get(result.size() - 1).markAsDone();
+                    }
+                //[E][ ] project meeting (from: Mon 2pm to: Fri 4pm)
+                } else if (line.startsWith("[E]")) {
+                    int fromIndex = line.indexOf(" (from: ");
+                    int toIndex = line.indexOf(" to: ");
+                    result.addTask(new Event(
+                        line.substring(7, fromIndex), 
+                        line.substring(fromIndex + 8, toIndex),
+                        line.substring(toIndex + 5, line.length() - 1)),
```

Comment:
I like what you did here to improve the readability of your code as the readability is affected by the multiple nesting. Thanks for the inspiration!